File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
  if (preorder.length == 0) return null;
  let rootIdxInInorder = getIndex(inorder, preorder[0])

  let left = buildTree(
    preorder.slice(1, rootIdxInInorder + 1),
    inorder.slice(0, rootIdxInInorder)
  );

  let right = buildTree(
    preorder.slice(rootIdxInInorder + 1),
    inorder.slice(rootIdxInInorder + 1)
  );
  return new TreeNode(preorder[0], left, right);
}
function getIndex(inorder, val) {
    let i = 0;
    while(val !==inorder[i]) {
          i++;
    }
    return i;
}

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder:
            return None
            
        indexroot = self.getIndex(inorder, preorder[0])
        
        leftnode = self.buildTree(preorder[1:indexroot + 1], inorder[0:indexroot])
        rightnode = self.buildTree(preorder[indexroot+1:], inorder[indexroot+1:])
        
        return TreeNode(preorder[0], leftnode, rightnode)
    
    def getIndex(self, inorder, node):
        for i, v in enumerate(inorder):
            if v == node:
                return i
        return 0